### PR TITLE
Unit tests for memory

### PR DESCRIPTION
**Pull request by KarolinePlum**
Issue #31

### CODE REVIEW

File: src/machine/state/mem.rs [hunk 1]
```diff
@@ -198,3 +198,78 @@ impl IndexMut<OctaAt> for Memory {
         self.buf.index_mut(pos)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn it_works() {
```
Comment (LukasKalbertodt) [inline src/machine/state/mem.rs]: Maybe you forgot to change this name to a non dummy one?

**Comment by teiesti** [comment]
Merges like 1a4efcf should not happen since they make the history very hard to read. A rebase is better. You can correct that with `git reset`, `git cherry-pick` and `git push -f` but be very careful with these commands because they can destroy data.

**Comment by LukasKalbertodt** [comment]
AFAIK, a rebase will sometimes remove merge commits from the history. So just try to rebase first, before using more dangerous commands ;-)

File: src/machine/state/mem.rs [hunk 2]
```diff
@@ -2,16 +2,16 @@ use machine::state::types::*;
 
 use std::ops::{Index, IndexMut};
 
-#[derive(Copy, Clone)]
+#[derive(Copy, Clone, Debug, PartialEq)]
```
Comment (teiesti) [inline src/machine/state/mem.rs:5]: Why do we need Debug?
Why PartialEq? Why not Eq?

File: src/machine/state/mem.rs [hunk 3]
```diff
@@ -198,3 +198,95 @@ impl IndexMut<OctaAt> for Memory {
         self.buf.index_mut(pos)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn mem_test() {
```
Comment (teiesti) [inline src/machine/state/mem.rs:207]: It think, you combine more than one test within this function! I suggest to split them!

File: src/machine/state/mem.rs [hunk 4]
```diff
@@ -198,3 +198,95 @@ impl IndexMut<OctaAt> for Memory {
...
+        assert_eq!(mem[OctaAt(0)], 0x60000u64.into());
+
+    }
+
+    #[test]
+    fn equivalent_indices() {
+
+        let mut mem = Memory::with_capacity(15);
+
+        for i in 0 .. 16 {
```
Comment (teiesti) [inline src/machine/state/mem.rs:251]:
```rust
for i in 0..16
```

**Review by teiesti — CHANGES_REQUESTED** [review]

**Comment by teiesti** [comment]
I think we need tests that check if 

1. the memory is initialized with the right capacity:
   - oversized (too large)
   - undersized (too small)
   - allocated in chunks of 64 bits
   - initialized with zeros
2. the index operators work correct
   - no "out of octa" read or write
   - byte order (endianess)
   - esp.: What happens if s.o. write e.g. a Octa first and then reads the Byte, Wyde, Tetra?
3. corner cases are implemented properly
   - "out of bounds" read/write should panic
   - reading or writing the first/last byte should work
4. lifetimes work correctly
   - one mutable reference XOR multiple immutable references

I think we cannot write unit tests, that check number 4. Can we?

Does anybody have more ideas what to test?

**Comment by jonas-schievink** [comment]
4 can only really be checked via a compile-fail test (a test that makes sure a piece of code is rejected by the compiler), so I suggest to skip it (I've eyeballed the implementation and it should be correct in this regard, you can also easily test manually). If you really want to write such a test then you can use https://github.com/laumann/compiletest-rs

**Comment by teiesti** [comment]
There is a new pull-request replacing this one. Therefore: Closed.

**Pull request closed without merging**